File: week1/credit/credit.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <cs50.h>

bool check_sum(long long id);
char* get_type(long long id);

int main(void) {
    printf("Number: ");
    long long id = get_long_long();
    char* type = get_type(id);

    if(strcmp(type, "INVALID") != 0) {
        bool isValid = check_sum(id);
        if (isValid) {
            printf("%s\n", type);
        }
    }
    else {
        printf("INVALID\n");
    }
}

bool check_sum(long long id) {
    int size = log10(id) + 1;
    long long x = 10;
    int sum = 0;

    for (int i = 0; i < size; i++) {
        int n = 0;
        if (i == 0) {
            n = (id % x);
        }
        else {
            n = (id % x) / (x / 10);
        }
        if (i % 2 == 1) {
            n = n * 2;
            if (n >= 10) {
                n = (n % 10) + 1;
            }
        }
        sum = sum + n;
        x = x * 10;
    }
    return sum % 10 == 0;
}

char* get_type(long long id) {
    int size = log10(id) + 1;
    int typeNumber = id / pow(10, size - 2);
    if ((typeNumber / 10) == 4) {
        return "VISA";
    }
    switch (typeNumber) {
        case 34:
        case 37:
            return "AMEX";
        case 51:
        case 52:
        case 53:
        case 54:
        case 55:
            return "MASTERCARD";
    }
    return "INVALID";
}

File: week8/credit/credit.py
def main():
    id = input("Number: ")
    card_type = get_type(id);
    if card_type != "INVALID":
        isValid = check_sum(id)
        if isValid:
            print(card_type)
        else:
            print("INVALID")
    else:
        print("INVALID")

def check_sum(id):
    sum = 0
    for i in range(len(id) - 1, -1, -1):
        n = int(id[i])
        if i % 2 == 1:
            n *= 2
            if n > 10:
                n = (n % 10) + 1
        sum += n
    return sum % 10 == 0

def get_type(id):
    AMEX_PREFIX = ["34", "37"]
    MASTERCARD_PREFIX = ["51", "52", "53", "54", "55"]
    if id[:1] == "4":
        return "VISA"
    if id[:2] in AMEX_PREFIX:
        return "AMEX"
    if id[:2] in MASTERCARD_PREFIX:
        return "MASTERCARD"
    return "INVALID";

if __name__ == "__main__":
    main()